python/ray/data/_internal/execution/operators/map_operator.py

```diff
@@ -169,13 +191,86 @@ def __init__(
         # Callback functions that generate additional task kwargs
         # for the map task.
         self._map_task_kwargs_fns: List[Callable[[], Dict[str, Any]]] = []
+        # Callback for when first input bundle is ready (before task submission).
+        # Receives schema from the first bundle for deferred initialization
+        # (e.g., schema evolution for Iceberg writes via on_write_start).
+        self._on_start: Optional[Callable[[Optional["pa.Schema"]], None]] = on_start
+        self._on_start_called = False
+        # _map_transformer_ref is lazily initialized on first access.
+        # This ensures on_start callback (if registered) can modify the transformer
+        # before serialization (e.g., for Iceberg schema evolution).
+        self.__map_transformer_ref = None
+
+    @property
+    def _map_transformer_ref(self):
+        """Lazily serialize _map_transformer to object store on first access.
+
+        Deferred until first task submission so that on_start callbacks
+        (e.g., on_write_start for Iceberg) can modify the transformer state
+        before serialization.
+        """
+        if self.__map_transformer_ref is None:
+            self.__map_transformer_ref = ray.put(self._map_transformer)
+            self._warn_large_udf()
+        return self.__map_transformer_ref
 
     def add_map_task_kwargs_fn(self, map_task_kwargs_fn: Callable[[], Dict[str, Any]]):
         """Add a callback function that generates additional kwargs for the map tasks.
         In the map tasks, the kwargs can be accessible via `TaskContext.kwargs`.
         """
         self._map_task_kwargs_fns.append(map_task_kwargs_fn)
 
+    def _notify_first_input(self, bundled_input: RefBundle) -> None:
+        """Invoke on_start callback with schema if registered and not yet invoked.
+
+        Used for deferred initialization that needs schema from the first bundle
+        (e.g., schema evolution for Iceberg writes via on_write_start).
+        """
+        if not self._on_start_called and self._on_start is not None:
+            schema = self._get_schema_from_bundle(bundled_input)
+            self._on_start(schema)
+            self._on_start_called = True
+            # Note: _map_transformer_ref is lazily initialized, so no need to
+            # re-serialize here - it will be created with the updated state
+            # when first accessed in _add_bundled_input.
+
+    def _get_schema_from_bundle(self, bundle: RefBundle) -> Optional["pa.Schema"]:
+        """Extract PyArrow schema from a RefBundle.
+
+        For Arrow schemas, returns directly. For Pandas blocks, runs a lightweight
+        remote task to convert a 1-row sample to Arrow and extract the schema.
+        This ensures schema consistency with actual block conversion logic without
+        fetching block data to the driver.
+        """
+        import pyarrow as pa
```

Comment:
Why is this an Operator's method (we're not using self, there's no overriding in subclasses, etc)